tests/custom_lookups/tests.py

```diff
@@ -647,3 +671,61 @@ def test_subquery_usage(self):
                 id__in=Author.objects.filter(age__div3=2)
             )
             self.assertSequenceEqual(qs, [a2])
+
+
+class RegisterLookupTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.obj1 = Company.objects.create(name="apple", place="france")
+        cls.obj2 = Company.objects.create(name="atom", place="bangladesh")
+        cls.obj3 = Company.objects.create(name="cherry", place="finland")
+        cls.obj4 = Company.objects.create(name="all", place="fig")
+
+    def test_register_unregister_instance_lookup(self):
+        with register_lookup(Company.name.field, CustomStartsWith):
+            self.assertEqual(
+                Company.name.field.instance_lookups, {"sw": CustomStartsWith}
+            )
+        self.assertEqual(Company.name.field.instance_lookups, {})
+
+    def test_get_lookup_same_field(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertEqual(Company.name.field.get_lookup("sw"), CustomStartsWith)
+            self.assertEqual(Company.place.field.get_lookup("sw"), CustomStartsWith)
+
+    def test_get_lookup_specific_instance(self):
+        with register_lookup(Company.place.field, CustomStartsWith):
+            self.assertEqual(Company.name.field.get_lookup("sw"), None)
+            self.assertEqual(Company.place.field.get_lookup("sw"), CustomStartsWith)
+
+    def test_class_lookup_with_filter(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertSequenceEqual(
+                Company.objects.filter(name__sw="a"), [self.obj1, self.obj2, self.obj4]
+            )
+            self.assertSequenceEqual(Company.objects.filter(name__sw="c"), [self.obj3])
+
+    def test_instance_lookup_with_filter(self):
+        with register_lookup(Company.name.field, StartsWith4Char):
+            self.assertSequenceEqual(
+                Company.objects.filter(name__swf="a"), [self.obj1, self.obj2]
+            )
+            self.assertSequenceEqual(Company.objects.filter(name__swf="c"), [self.obj3])
+            with self.assertRaises(FieldError):
+                Company.objects.filter(place__swf="c")
+
+    def test_instance_overide_class_lookups_with_filter(self):
+        with register_lookup(
+            models.CharField, CustomStartsWith, lookup_name="start"
+        ), register_lookup(Company.place.field, StartsWith4Char, lookup_name="start"):
+            self.assertSequenceEqual(
+                Company.objects.filter(place__start="f"),
+                [
+                    self.obj1,
+                    self.obj3,
+                ],
+            )
+            self.assertSequenceEqual(
+                Company.objects.filter(name__start="a"),
+                [self.obj1, self.obj2, self.obj4],
+            )
```

Comment:
Can we test this with nested context processors to see that unregistering also works? e.g.
```python
with register_lookup(
    models.CharField, CustomStartsWith, lookup_name="start")
):
    with register_lookup(Company.place.field, StartsWith4Char, lookup_name="start"):
        self.assertCountEqual(
            Company.objects.filter(place__start="f"),
            [self.obj1, self.obj3],
        )
        self.assertCountEqual(
            Company.objects.filter(name__start="a"),
            [self.obj1, self.obj2, self.obj4],
        )
    self.assertCountEqual(
        Company.objects.filter(place__start="f"),
        [self.obj1, self.obj3, self.obj4],
    )
```

Reply:
Done👍